### PR TITLE
migrate cupertino to nullsafety

### PR DESCRIPTION
**Pull request by a14n**
## Description

NNBD migration for last files of cupertino

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/nav_bar.dart [hunk 1]
```diff
@@ -1451,12 +1449,12 @@ class _BackLabel extends StatelessWidget {
   Widget build(BuildContext context) {
     if (specifiedPreviousTitle != null) {
       return _buildPreviousTitleWidget(context, specifiedPreviousTitle, null);
-    } else if (route is CupertinoRouteTransitionMixin<dynamic> && !route.isFirst) {
+    } else if (route is CupertinoRouteTransitionMixin<dynamic> && !route!.isFirst) {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/nav_bar.dart:1452]: Shouldn't the compiler assume that `route` is non-nullable after the `is` comparing it to a non-nullable type?

File: packages/flutter/lib/src/cupertino/nav_bar.dart [hunk 2]
```diff
@@ -1514,7 +1512,7 @@ class _TransitionableNavigationBar extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     assert(() {
-      bool inHero;
+      bool? inHero;
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/nav_bar.dart:1515]: Can't this be non-nullable and default to false?

File: packages/flutter/lib/src/cupertino/tab_scaffold.dart [hunk 3]
```diff
@@ -417,12 +415,12 @@ class _CupertinoTabScaffoldState extends State<CupertinoTabScaffold> {
...
               child: widget.tabBar.copyWith(
-                currentIndex: _controller.index,
+                currentIndex: _controller!.index,
                 onTap: (int newIndex) {
-                  _controller.index = newIndex;
+                  _controller!.index = newIndex;
                   // Chain the user's original callback.
                   if (widget.tabBar.onTap != null)
-                    widget.tabBar.onTap(newIndex);
+                    widget.tabBar.onTap!(newIndex);
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/tab_scaffold.dart:423]: Could this be on one line?
```dart
widget.tabBar.onTap?.call(newIndex);
```

File: packages/flutter/lib/src/cupertino/text_selection.dart [hunk 4]
```diff
@@ -392,18 +390,18 @@ class _ToolbarRenderBox extends RenderShiftedBox {
...
-      ..strokeWidth = 2.0
-      ..style = PaintingStyle.stroke;
-
-      final _ToolbarParentData childParentData = child.parentData as _ToolbarParentData;
-      context.canvas.drawPath(_clipPath().shift(offset + childParentData.offset), _debugPaint);
+        ..shader = ui.Gradient.linear(
+          const Offset(0.0, 0.0),
+          const Offset(10.0, 10.0),
+          <Color>[const Color(0x00000000), const Color(0xFFFF00FF), const Color(0xFFFF00FF), const Color(0x00000000)],
+          <double>[0.25, 0.25, 0.75, 0.75],
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/text_selection.dart:397]:
```suggestion
		  const <Color>[Color(0x00000000), Color(0xFFFF00FF), Color(0xFFFF00FF), Color(0x00000000)],
		  const <double>[0.25, 0.25, 0.75, 0.75],
```

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/cupertino/nav_bar.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/nav_bar.dart:1452]: promotion to non-null does not work for fields/getters. See dart-lang/language#1187 for possible improvement.

Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/nav_bar.dart:1452]: Darn.  OK.

File: packages/flutter/lib/src/cupertino/text_selection.dart (hunk 4 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/text_selection.dart:397]: Done

File: packages/flutter/lib/src/cupertino/tab_scaffold.dart (hunk 3 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/tab_scaffold.dart]: Definitively. It's a matter of style. Done.

File: packages/flutter/lib/src/cupertino/nav_bar.dart (hunk 2 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/nav_bar.dart]: Done

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Pull request merged**